### DISCUSSION TITLE
Z80 NMOS 8MHz instead of Z84C0020?

### DISCUSSION THREAD

**Discussion by LuckStilon** [UzixLS/zx-sizif-512#65] [General]
For some time now it has become increasingly difficult to buy the Z84C0020 (**CMOS 20MHz**) processor. ZiLOG has ended their production, and the Chinese are massively cheating: they are repainting **Z80 NMOS 8MHz** processors to Z84C0020. Out of 10 bought from them, 7 are cheated, and this proportion increases over time because people returns them fake chips. **Is it possible to modify the project for these processors?** I mean using NMOS instead of CMOS and limit the maximum speed to 7 MHz.

**Comment by UzixLS** [comment]
Great work around NMOS Z80 has already been done in commit 8e092541dd402329574b6fbecd31f13e869f7e80 (see also https://t.me/uzixls/92 and https://t.me/uzixls/93), but further investigation is needed as this commit breaks compatibility with some SRAM chips.

**Reply by LuckStilon** [reply]
Oh, great news! I see you did a big job! I'm still using the 2022-10-02 version, which is probably why I still had the problem. I have a few dozen NMOS processors. I'll try to test them on the latest version and come back with statistics. Luckily, I use only one version of SRAM, from one reliable source (tme.eu). Here's my test bench :-) Thanks!

![Clipboard01](https://github.com/user-attachments/assets/15f83a4b-6d17-400c-be15-d5fa70d3be10)

**Reply by UzixLS** [reply]
New version isn't released for now as it requires further improvement.

**Reply by LuckStilon** [reply]
I have an "alpha" pof and rom files, version dated 2024-01-28 from somewhere. Should be enough for testing.

Additionally - is it possibly to add parameter "**CPU type**" (NMOS/CMOS) to **MAGIC->Advanced** menu? Switching to NMOS would limit the processor speed to 7MHz (instead of 14 MHz) during SD card operations in **Auto CPU - ON** mode, and limit the speed selection to a maximum of 7MHz. If I remember correctly, the fastest NMOS processors were manufactured for 8MHz.

**Reply by UzixLS** [reply]
I don't like idea to have 9999 configuration options in menu, but it's possible to autodetect NMOS CPU at boot time and automatically limit speed for "auto cpu freq" option.

**Comment by UzixLS** [comment]
CPU type autodetection was implemented in commit 23445b482e61d14f514820286fc7b03e76e1094c and ll be included in the next release.

**Reply by LuckStilon** [reply]
Great news! Some time ago I tested over 25 NMOS processors and modified the ROM for my needs: removed the autoturbo parameter and the 14MHz speed.

Are there any things in the current version of the code that are known to work incorrectly and needs to be fixed?

![obraz](https://github.com/user-attachments/assets/b2a64e8c-f04d-4407-8019-fc429b065668)

**Reply by UzixLS** [reply]
There are known issues after commit 8e092541dd402329574b6fbecd31f13e869f7e80 - it's breaks compatibility with some bootleg SRAM chips, which works almost fine with previous version.

**Reply by arleybls** [reply]
> CPU type autodetection was implemented in commit [23445b4](https://github.com/UzixLS/zx-sizif-512/commit/23445b482e61d14f514820286fc7b03e76e1094c) and ll be included in the next release.

Does the soundcard still works with nmos at 7mhz?

**Comment by LuckStilon** [comment]
Just tested current "trunk" version. I compiled current ROM and POF files. Found two problems: 
1). Some problems with joystick in Kempston mode. About about every second-third case after restart, in the MAGIC menu the joystick does not work. The same in Retroleum DiagROM: on the keyboard test, the joystick emulates key presses 6,7,8,9,0. Reading in BASIC (print out 31) in this fail-mode does not work either.
2). Z80 at speed > 3.5MHz works unevenly. It doesn't matter much, but you can hear it in the music, for example, after starting the game Starquake. The music plays fast (due to fast clock speed), but not cleanly, with glissando. On the last released version 2023-08-20 there was no such problem - the music played faster, but without the out of tune and glissandos.

The above problems occur on every type of processor.

Other than that is OK. Autodetecting and changing the processor frequency works great. Tested on original, new Z84C0020, and few chinese Z80 CMOS 20MHz and NMOS 8MHz.

**Reply by AtomicRPM** [reply]
I have a question, are you simply running make to build?  I rub make in rom_src, and them run make in rom and the .bin I write to rom freezes with a blue border and no menu appears when the menu button is held down.

**Reply by AtomicRPM** [reply]
Resolved.
